modules/dop/providers/cms/cicdcmsservice.go

```diff
@@ -127,6 +131,48 @@ func (s *CICDCmsService) CICDCmsCreateOrUpdate(ctx context.Context, req *CICDCms
 		}))
 	}()
 
+	if req.Batch {
+		// get pre configs
+		newContext := transport.WithHeader(ctx, metadata.New(map[string]string{httputil.InternalHeader: "true"}))
+		preCMS, err := s.p.PipelineCms.GetCmsNsConfigs(newContext, &cmspb.CmsNsConfigsGetRequest{
+			PipelineSource: updateRequest.PipelineSource,
+			Ns:             updateRequest.Ns,
+		})
+		if err != nil {
+			return false, err
+		}
+
+		// get unnecessary keys
+		var deleteKeys []string
+		for _, v := range preCMS.Data {
+			if v.Type == "FILE" {
+				continue
+			}
+
+			var find = false
+			for _, conf := range keys {
+				if conf != v.Key {
+					continue
+				}
+				find = true
+			}
+			if !find {
+				deleteKeys = append(deleteKeys, v.Key)
+			}
+		}
+
+		// Delete unnecessary keys in the project pipeline configs
+		_, err = s.p.PipelineCms.DeleteCmsNsConfigs(newContext, &cmspb.CmsNsConfigsDeleteRequest{
+			Ns:             updateRequest.Ns,
+			PipelineSource: updateRequest.PipelineSource,
+			DeleteKeys:     deleteKeys,
+		})
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// create or update configs
```

Comment:
what will happen if delete success but update failed?

Reply:
The changed key will be lost